From 1e2b9eed437b86950450328fea28762443fa1336 Mon Sep 17 00:00:00 2001
From: Andrew Kelley <andrew@ziglang.org>
Date: Tue, 23 Sep 2025 16:37:53 -0700
Subject: [PATCH 113/133] forbid trivial local address returned from functions
 (#25333)
X-Developer-Signature: v=1; a=openpgp-sha256; l=3992; i=xtex@astrafall.org;
 h=from:subject; bh=TZloIxSXfP2Na6RKwIoqefTkQIr6M7CviYBsO9EBtMw=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78nhVDx1s3rHMpLn0yfYv5TV/en7/2Hu13JFH3Xu7+
 9oXeZyxHRNZGMS4GCzFFFmKDBu8WXXS+UWXlcvCzGFlAhkiLdLAwMDAwMLAl5uYV2qkY6Rnqm2o
 Z2ioY6RjxMDFKQBTHbKKkeHqGyMbdb3CFtsAvvi+s6+XmkYu1LWeuYFHT+NO0Jkv7+IY/ldNnul
 7zf3x9geas74V/Ppbwet6eyPn5xp3p46Es04fPzMCAA==
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

progress towards #25312

(cherry picked from commit ba137783ed4152ebe8d7b98dca855c551fc924a7)
Manually merged
Change-Id: Iab8ab5e798472bc0572658cbe0dc4fba5d5ac428
---
 lib/std/zig/AstGen.zig                        | 31 ++++++++++++++-----
 .../compile_errors/return_addr_of_local.zig   |  9 ++++++
 2 files changed, 32 insertions(+), 8 deletions(-)
 create mode 100644 test/cases/compile_errors/return_addr_of_local.zig

diff --git a/lib/std/zig/AstGen.zig b/lib/std/zig/AstGen.zig
index da93086396b2..1e4481ce5eec 100644
--- a/lib/std/zig/AstGen.zig
+++ b/lib/std/zig/AstGen.zig
@@ -383,6 +383,8 @@ const ResultInfo = struct {
         assignment,
         /// No specific operator in particular.
         none,
+        /// The expression is operand to address-of which is the operand to a return expression.
+        return_addrof,
     };
 };
 
@@ -955,7 +957,14 @@ fn expr(gz: *GenZir, scope: *Scope, ri: ResultInfo, node: Ast.Node.Index) InnerE
                 _ = try gz.addUnTok(.validate_ref_ty, res_ty_inst, tree.firstToken(node));
                 break :rl .{ .ref_coerced_ty = res_ty_inst };
             } else .ref;
-            const result = try expr(gz, scope, .{ .rl = operand_rl }, tree.nodeData(node).node);
+            const operand_node = tree.nodeData(node).node;
+            const result = try expr(gz, scope, .{
+                .rl = operand_rl,
+                .ctx = switch (ri.ctx) {
+                    .@"return" => .return_addrof,
+                    else => .none,
+                },
+            }, operand_node);
             return rvalue(gz, ri, result, node);
         },
         .optional_type => {
@@ -8420,13 +8429,19 @@ fn localVarRef(
                     local_ptr.used = .fromToken(ident_token);
                 }
 
-                // Can't close over a runtime variable
-                if (num_namespaces_out != 0 and !local_ptr.maybe_comptime and !gz.is_typeof) {
-                    const ident_name = try astgen.identifierTokenString(ident_token);
-                    return astgen.failNodeNotes(ident, "mutable '{s}' not accessible from here", .{ident_name}, &.{
-                        try astgen.errNoteTok(local_ptr.token_src, "declared mutable here", .{}),
-                        try astgen.errNoteNode(capturing_namespace.node, "crosses namespace boundary here", .{}),
-                    });
+                if (!local_ptr.maybe_comptime and !gz.is_typeof) {
+                    if (num_namespaces_out != 0) {
+                        const ident_name = try astgen.identifierTokenString(ident_token);
+                        return astgen.failNodeNotes(ident, "mutable '{s}' not accessible from here", .{ident_name}, &.{
+                            try astgen.errNoteTok(local_ptr.token_src, "declared mutable here", .{}),
+                            try astgen.errNoteNode(capturing_namespace.node, "crosses namespace boundary here", .{}),
+                        });
+                    } else if (ri.ctx == .return_addrof) {
+                        const ident_name = try astgen.identifierTokenString(ident_token);
+                        return astgen.failNodeNotes(ident, "returning address of expired local variable '{s}'", .{ident_name}, &.{
+                            try astgen.errNoteTok(local_ptr.token_src, "declared runtime-known here", .{}),
+                        });
+                    }
                 }
 
                 switch (ri.rl) {
diff --git a/test/cases/compile_errors/return_addr_of_local.zig b/test/cases/compile_errors/return_addr_of_local.zig
new file mode 100644
index 000000000000..1f29f39cdd01
--- /dev/null
+++ b/test/cases/compile_errors/return_addr_of_local.zig
@@ -0,0 +1,9 @@
+export fn b() void {
+    var x: i32 = 1234;
+    return &x;
+}
+
+// error
+//
+// :3:13: error: returning address of expired local variable 'x'
+// :2:9: note: declared runtime-known here
-- 
2.52.0

